### PR TITLE
Switches embedded server from Tomcat to Undertow

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This switches from Tomcat to Undertow, then MVC to raw Undertow handler to process http messages (for apps sending data directly to zipkin).

The test uses a real backend even if local elasticsearch. The test results in a lot more data than my local ES can swallow, so it is also a test of how fast we can drop data, and how well our rate-limiting works.

So, after this change, the same zipkin server against local elasticsearch can swallow 3x more traffic than before.

The below results were a best of 5 for before and after.. tests were interleaved to ensure elasticsearch wasn't skewing.

```
# before
Running 30s test @ http://localhost:9411/api/v2/spans
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    14.50ms   37.19ms 527.63ms   93.43%
    Req/Sec     2.02k     0.94k    6.66k    71.75%
  Latency Distribution
     50%    4.20ms
     75%    8.61ms
     90%   24.53ms
     99%  198.98ms
  212147 requests in 30.05s, 24.92MB read
  Socket errors: connect 0, read 1, write 0, timeout 19
Requests/sec:   7058.96
Transfer/sec:    849.19KB

# after
Running 30s test @ http://localhost:9411/api/v2/spans
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    16.28ms   33.11ms 388.07ms   90.37%
    Req/Sec     6.43k     2.86k   14.55k    64.65%
  Latency Distribution
     50%    3.03ms
     75%   14.59ms
     90%   47.73ms
     99%  155.01ms
  761153 requests in 30.05s, 76.22MB read
Requests/sec:  25332.61
Transfer/sec:      2.54MB
```

Here's the command which starts a server and uses elasticsearch local storage:
```bash
$ STORAGE_TYPE=elasticsearch java -jar ./zipkin-server/target/zipkin-server-*-exec.jar
$ wrk -t4 -c128 -d30s http://localhost:9411/api/v2/spans -s zipkin-json.lua --latency
```

Here's the script:
```lua
wrk.method = "POST"
wrk.headers["Content-Type"] = "application/json"
wrk.body   = '[{"traceId":"4d1e00c0db9010db86154a4ba6e91385","parentId":"86154a4ba6e91385","id":"4d1e00c0db9010db","kind":"CLIENT","name":"get","timestamp":1472470996199000,"duration":207000,"localEndpoint":{"serviceName":"frontend","ipv6":"7::0.128.128.127"},"remoteEndpoint":{"serviceName":"backend","ipv4":"192.168.99.101","port":9000},"annotations":[{"timestamp":1472470996238000,"value":"foo"},{"timestamp":1472470996403000,"value":"bar"}],"tags":{"http.path":"/api","clnt/finagle.version":"6.45.0"}}]'
```

I looked at the jar, and it is roughly the same size as before. Since
undertow uses the relatively obscure xnio library, it isn't likely to
conflict with more often used things like netty.

### CODE REVIEW

**Comment by adriancole** [comment]
best run of several (locally) with tomcat:
```
Running 30s test @ http://localhost:9411/api/v2/spans
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    11.23ms   18.08ms 400.13ms   93.27%
    Req/Sec     4.23k   496.83     7.75k    77.38%
  Latency Distribution
     50%    7.10ms
     75%    8.86ms
     90%   16.97ms
     99%  100.63ms
  504453 requests in 30.05s, 59.26MB read
  Socket errors: connect 0, read 11, write 0, timeout 0
Requests/sec:  16789.47
Transfer/sec:      1.97MB
```

Best run of several (locally) with undertow
```
Running 30s test @ http://localhost:9411/api/v2/spans
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     9.79ms   16.26ms 421.10ms   93.06%
    Req/Sec     5.05k   547.20     8.77k    73.58%
  Latency Distribution
     50%    5.51ms
     75%    7.37ms
     90%   17.48ms
     99%   87.97ms
  604521 requests in 30.10s, 89.36MB read
Requests/sec:  20085.30
Transfer/sec:      2.97MB
```

**Comment by codefromthecrypt** [comment]
cc @anuraaga @shakuzen @denyska FYI a 3x bump in http ingest throughput

**Comment by jorgheymans** [comment]
👍

**Comment by adriancole** [comment]
I'm going to revise this to yank the MVC controller as there's a large amount of latency added there. Here's raw undertow (still wired with spring boot), which has dramatically better latency characteristics. The impact is we can chop off several milliseconds off the better requests and an order of magnitude off the worse ones. This is particularly helpful when collector-tier sampling is in place as it means we can skip data much quicker under a surge.

```bash
$ wrk -t4 -c128 -d30s http://localhost:9411/api/v2/spans -s zipkin-json.lua --latency
Running 30s test @ http://localhost:9411/api/v2/spans
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    11.23ms   18.08ms 400.13ms   93.27%
    Req/Sec     4.23k   496.83     7.75k    77.38%
  Latency Distribution
     50%    7.10ms
     75%    8.86ms
     90%   16.97ms
     99%  100.63ms
  504453 requests in 30.05s, 59.26MB read
  Socket errors: connect 0, read 11, write 0, timeout 0
Requests/sec:  16789.47
Transfer/sec:      1.97MB
```

**Review by denyska — APPROVED** [review]

**Comment by adriancole** [comment]
ok code is up. I changed the test to use a real backend even if local elasticsearch. The test results in a lot more data than ES can swallow, so it is also a test of how fast we can drop data, and how well our rate-limiting works.

So, after this change, the same zipkin server against local elasticsearch can swallow 3x more traffic than before.

The below results were a best of 5 for before and after.. tests were interleaved to ensure elasticsearch wasn't skewing.

```
# before
Running 30s test @ http://localhost:9411/api/v2/spans
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    14.50ms   37.19ms 527.63ms   93.43%
    Req/Sec     2.02k     0.94k    6.66k    71.75%
  Latency Distribution
     50%    4.20ms
     75%    8.61ms
     90%   24.53ms
     99%  198.98ms
  212147 requests in 30.05s, 24.92MB read
  Socket errors: connect 0, read 1, write 0, timeout 19
Requests/sec:   7058.96
Transfer/sec:    849.19KB

# after
Running 30s test @ http://localhost:9411/api/v2/spans
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    16.28ms   33.11ms 388.07ms   90.37%
    Req/Sec     6.43k     2.86k   14.55k    64.65%
  Latency Distribution
     50%    3.03ms
     75%   14.59ms
     90%   47.73ms
     99%  155.01ms
  761153 requests in 30.05s, 76.22MB read
Requests/sec:  25332.61
Transfer/sec:      2.54MB
```

**Comment by codefromthecrypt** [comment]
PS there's clearly more possible here, just looking for a cheap win

**Comment by codefromthecrypt** [comment]
ok I rejigged the cors filter into its own commit (tested it with the zipkin-js browser example). I'm going to test a proxied zipkin ui (nginx) before merging

**Comment by codefromthecrypt** [comment]
noticed our metrics were double-counting so fixed that before proceeding. I want to watch our prometheus dashboard before and after..

**Comment by codefromthecrypt** [comment]
yanked the servlet filter for request duration as we don't need it anymore

**Comment by codefromthecrypt** [comment]
TL;DR; this is without a doubt far better especially in my container environment.

I tested in docker. This version does quite a lot better in terms of resources used, time in GC, etc.. Using elasticsearch5 image and grafana etc

Before: best is 587 requests/second with P99 1second
```
Running 30s test @ http://192.168.99.100:9411/api/v2/spans
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   251.37ms  223.66ms   1.96s    70.74%
    Req/Sec   148.22     90.62   540.00     66.30%
  Latency Distribution
     50%  210.87ms
     75%  357.93ms
     90%  552.59ms
     99%  995.82ms
  17665 requests in 30.06s, 2.85MB read
Requests/sec:    587.62
Transfer/sec:     97.06KB
```

After: best is 2.5K requests/second with P99 190ms
```
Running 30s test @ http://192.168.99.100:9411/api/v2/spans
  4 threads and 128 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    53.76ms   35.06ms 346.09ms   79.44%
    Req/Sec   640.58    220.66     1.73k    75.34%
  Latency Distribution
     50%   48.23ms
     75%   64.04ms
     90%   90.27ms
     99%  191.56ms
  76292 requests in 30.01s, 10.99MB read
Requests/sec:   2541.89
Transfer/sec:    374.83KB
```

All things monitored in graphite were measurably better including load, thread count, time in GC

screen shot 2017-11-23 at 10 41 43 pm
screen shot 2017-11-23 at 10 47 40 pm
screen shot 2017-11-23 at 10 41 34 pm
screen shot 2017-11-23 at 10 47 25 pm
screen shot 2017-11-23 at 10 41 28 pm
screen shot 2017-11-23 at 10 47 31 pm
